pkg/skaffold/runner/dev.go

```diff
@@ -42,43 +42,48 @@ func (r *SkaffoldRunner) Dev(ctx context.Context, out io.Writer, artifacts []*la
 	defer forwarderManager.Stop()
 
 	// Create watcher and register artifacts to build current state of files.
-	changed := changes{}
 	onChange := func() error {
-		defer changed.reset()
+		defer r.changeSet.reset()
 
 		logger.Mute()
 
-		for _, a := range changed.dirtyArtifacts {
+		for _, a := range r.changeSet.dirtyArtifacts {
 			s, err := sync.NewItem(a.artifact, a.events, r.builds, r.runCtx.InsecureRegistries)
 			if err != nil {
 				return errors.Wrap(err, "sync")
 			}
 			if s != nil {
-				changed.AddResync(s)
+				r.changeSet.AddResync(s)
 			} else {
-				changed.AddRebuild(a.artifact)
+				r.changeSet.AddRebuild(a.artifact)
 			}
 		}
 
+		if r.changeSet.needsReload || len(r.changeSet.needsRebuild) > 0 || r.changeSet.needsRedeploy {
+			// if any action is going to be performed, reset the monitor's changed component tracker for debouncing
+			defer r.Trigger.WatchForChanges(out)
```

Comment:
since `devLoop()` is happening every time a trigger is received (even if it's a noop), this won't work. we need to only call this function whenever we've actually performed actions in the dev loop, not just ever time we get in there.

reworked this to embed the call in the listener.